server/src/main/java/com/thoughtworks/go/config/ConfigReposMaterialParseResultManager.java

```diff
@@ -33,81 +30,90 @@
-
+        //config repository was parsed, but does not have merge or clone related errors.
+        if (result != null && result.isSuccessful()) {
             HealthStateScope healthStateScope = HealthStateScope.forPartialConfigRepo(fingerprint);
             List<ServerHealthState> serverHealthStates = serverHealthService.filterByScope(healthStateScope);
-            // It should be noted that the server health state and the result can be thought of as the following:
-            // serverHealthState is the current state
-            // result.isSuccessfully() is the state at the last parse
-
-            // here we're checking that the current state is bad, but the previous state was good
-            if (!serverHealthStates.isEmpty() && result.isSuccessful()) {
-                result.setException(represent(serverHealthStates.get(0)));
+            if (!serverHealthStates.isEmpty()) {
+                result.setException(asError(serverHealthStates.get(0)));
 
                 //clear out the good modification, in case good modification is same as of latest parsed modification
                 if (result.getLatestParsedModification().equals(result.getGoodModification())) {
                     result.setGoodModification(null);
                     result.setPartialConfig(null);
                 }
-            // here we're checking that the current state is good, but the previous state was bad
-            } else if (serverHealthStates.isEmpty() && !result.isSuccessful()) {
-                // The issues that caused the previous state to be bad are gone.
-                // An example would be that previously a upstream pipeline didn't exist, but now does
-                // Clearing latest parsed modification to allow for another parse attempt
-                result.setLatestParsedModification(null);
             }
         }
 
         return result;
     }
 
-    private PartialConfigParseResult getMaterialResult(String fingerprint) {
-        HealthStateScope healthStateScope = HealthStateScope.forMaterialConfig(configRepoService.findByFingerprint(fingerprint).getMaterialConfig());
-        List<ServerHealthState> serverHealthStates = serverHealthService.filterByScope(healthStateScope);
-        if (!serverHealthStates.isEmpty()) {
-            return PartialConfigParseResult.parseFailed(null, represent(serverHealthStates.get(0)));
-        }
-        return null;
+    /**
+     * After a successful update of the server config, we should reparse any failed config-repos in case
+     * any upstream dependency issues have been resolved.
+     *
+     * @return a listener that clears the last parse states and errors related to config-repos so as to allow reparsing
+     */
+    ConfigChangedListener onConfigUpdateDoReparse() {
+        return newCruiseConfig -> {
+            for (Map.Entry<String, PartialConfigParseResult> entry : fingerprintOfPartialToParseResultMap.entrySet()) {
```

Comment:
This seems to clear the map on every config save, even ones where the config update does not involve changes to pipelines/config repos/scm materials. Is this intentional?

Can we possibly also use the `EntityConfigChangedListener` to improve this a bit so that config changes to pipelines/config repos/scm materials will selectively update just the result that needs updating?

The current code will have to remain as-is for handling a full config save when someone edits the XML.

Reply:
I was not confident that I knew all the ways merging a partial config could fail; missing template, pipeline, materials but might be others? This I made the criteria very broad.

I suppose at scale this could have performance implications that I do not know yet.

Given the different ways a merge can fail dependency checks, is there a good way to find which partialconfig result matches a config entity in order to selectively update affected results? Other than parsing the exception message, that is.